### PR TITLE
Add some user-defined parameters

### LINKED ISSUE #21: Add mag_normalization error
Currently, in MacroscopicEvolveM.cpp, mag_normalization_error=0.1 is hard-coded. This needs to be added as a user-defined inputs with appropriate docs.

### LINKED ISSUE #25: Change the M_max_iter and M_tol to user-defined input in the MacroscopicEvolveM_2nd.cpp
See the title.

### PR DESCRIPTION
**Pull request by ruohai0925**
1. Fix the issue #25 : change the M_max_iter and M_tol to user-defined input in the MacroscopicEvolveM_2nd.cpp; 

2. Fix the issue #21 : change the mag_normalized_error to user-defined input in the MacroscopicEvolveM.cpp; 

I validated all of these input variables. Here is the input file.

`################################
#################################
max_step = 1000
amr.n_cell = 8 8 8 # number of cells spanning the domain in each coordinate direction at level 0
amr.max_grid_size = 512 # maximum size of each AMReX box, used to decompose the domain
amr.blocking_factor = 8
geometry.coord_sys   = 0
geometry.is_periodic = 1  1  1
geometry.prob_lo     =  -1.5e-6 -1.5e-6 -1.5e-6
geometry.prob_hi     =   1.5e-6  1.5e-6  1.5e-6
amr.max_level = 0
#################################
#################################
warpx.verbose = 0
warpx.use_filter = 0
warpx.cfl = 4000
warpx.do_pml = 0

warpx.mag_time_scheme_order = 2 # default 1
**macroscopic.mag_normalized_error = 0.1 # default 0.1
macroscopic.mag_max_iter = 100 # default 100
macroscopic.mag_tol = 0.0001 # default 0.0001**

#algo.em_solver_medium = vacuum # vacuum/macroscopic
algo.em_solver_medium = macroscopic # vacuum/macroscopic
algo.macroscopic_sigma_method = laxwendroff # laxwendroff or backwardeuler
macroscopic.sigma_init_style = "parse_sigma_function" # parse or "constant"
macroscopic.sigma_function(x,y,z) = "0.0"
#macroscopic.sigma = 0.0
macroscopic.epsilon_init_style = "parse_epsilon_function" # parse or "constant"
macroscopic.epsilon_function(x,y,z) = "8.8541878128e-12"
#macroscopic.epsilon = 8.8541878128e-12
macroscopic.mu_init_style = "parse_mu_function" # parse or "constant"
macroscopic.mu_function(x,y,z) = "1.25663706212e-06"
#macroscopic.mu = 1.25663706212e-06
macroscopic.mag_Ms_init_style = "parse_mag_Ms_function" # parse or "constant"
macroscopic.mag_Ms_function(x,y,z) = "1.4e5"
#macroscopic.mag_Ms = 1.4e5 # in unit A/m, equal to 1750 Gauss
#unit conversion: 1 Gauss = (1000/4pi) A/m
macroscopic.mag_alpha_init_style = "parse_mag_alpha_function" 
macroscopic.mag_alpha_function(x,y,z) = "5e-01"
#macroscopic.mag_alpha = 5e-02 # alpha is unitless, typical values range from 1e-3 ~ 1e-5
macroscopic.mag_gamma_init_style = "parse_mag_gamma_function"
macroscopic.mag_gamma_function(x,y,z) = "-1.759e11"
#macroscopic.mag_gamma = 1.759e11 # gyromagnetic ratio is constant for electrons in all materials
particles.nspecies = 0
#################################
#################################
my_constants.pi = 3.14159265359
my_constants.L = 141.4213562373095e-6
my_constants.c = 299792458.
my_constants.wavelength = 1.2e-1
warpx.E_ext_grid_init_style = parse_E_ext_grid_function
warpx.Ez_external_grid_function(x,y,z) = 0.
warpx.Ex_external_grid_function(x,y,z) = 0.
warpx.Ey_external_grid_function(x,y,z) = 0.
#warpx.Ey_external_grid_function(x,y,z) = "1.e5*exp(-z**2/L**2)*cos(2*pi*z/wavelength)"
warpx.B_ext_grid_init_style = parse_B_ext_grid_function
#warpx.Bx_external_grid_function(x,y,z)= "-1.e5*exp(-z**2/L**2)*cos(2*pi*z/wavelength)/c"
warpx.Bx_external_grid_function(x,y,z)= 0.1759
#warpx.Bx_external_grid_function(x,y,z)= 0.
warpx.By_external_grid_function(x,y,z)= 0.
warpx.Bz_external_grid_function(x,y,z) = 0.
warpx.H_bias_ext_grid_init_style = parse_H_bias_ext_grid_function
warpx.Hx_bias_external_grid_function(x,y,z)= 0.
warpx.Hy_bias_external_grid_function(x,y,z)= 0.
warpx.Hz_bias_external_grid_function(x,y,z)= 3e4 # in A/m, equal to 382 Oersted
#warpx.Hz_bias_external_grid_function(x,y,z)= 0. # in A/m, equal to 382 Oersted
#unit conversion: 1 Gauss = 1 Oersted = (1000/4pi) A/m
#calculation of H_bias: H_bias (oe) = frequency / 2.8e6
#warpx.H_bias_ext_grid_init_style = constant
#warpx.Hx_bias_external_grid = 0
#warpx.Hy_bias_external_grid = 0
#warpx.Hz_bias_external_grid = 0
#warpx.M_ext_grid_init_style = parse_M_ext_grid_function
#warpx.Mx_external_grid_function(x,y,z)= "-1.e5*exp(-z**2/L**2)*cos(2*pi*z/wavelength)/c"
#warpx.My_external_grid_function(x,y,z)= 0.
#warpx.Mz_external_grid_function(x,y,z) = 0.
warpx.M_ext_grid_init_style = constant
warpx.M_external_grid = 140000. 0. 0.
#Diagnostics
diagnostics.diags_names = plt
plt.period = 1000
plt.diag_type = Full
#plt.fields_to_plot = Mx_xface My_xface Mz_xface Mx_yface My_yface Mz_yface Mx_zface My_zface Mz_zface
plt.fields_to_plot = Mx_xface My_xface Mz_xface
plt.plot_raw_fields = 0
`

### CODE REVIEW

**Review by ajnonaka — COMMENTED** [review]
I notice that sigma_init_style, epsilon_init_stype, mu_init_style, mag_normalized_error, mag_max_iter, and mag_tol do not have entries in WarpX/Docs/Source/running_cpp/parameters.rst.  Can we add descriptions here?

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H [hunk 1]
```diff
@@ -174,6 +178,16 @@ private:
      amrex::Real m_mag_alpha;
      /** gyromagnetic ratio, should be a negative value, only applies for magnetic materials */
      amrex::Real m_mag_gamma;
+
+     // normalization error of M field for checking, default 0.1
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H]:
```suggestion
     // If the magnitude of the magnetization deviates by more than this amount relative
     // to the user-defined Ms, abort.  Default 0.1.
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H [hunk 2]
```diff
@@ -174,6 +178,16 @@ private:
...
+     amrex::Real m_mag_normalized_error;
+
+     // maximum iteration time for the second-order time advancement scheme of M field, default 100
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H]:
```suggestion
     // maximum iteration count for the second-order time advancement scheme of M field, default 100
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp [hunk 3]
```diff
@@ -283,9 +283,9 @@ void FiniteDifferenceSolver::MacroscopicEvolveM_2nd (
 
         // initialize M_max_iter, M_iter, M_tol, M_iter_error
         // change these to user-defined input in the next PR
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp:285]:
```suggestion
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 4]
```diff
@@ -43,8 +43,9 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
         amrex::Real const dt,
         std::unique_ptr<MacroscopicProperties> const& macroscopic_properties )
     {
-    // Temporary value hard-coded for normalized error used for LLG.
-    amrex::Real mag_normalized_error = 0.1_rt; // normalization error of M field for checking
+
+        // normalization error of M field for checking
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]:
```suggestion
        // obtain the maximum relative amount we let M deviate from Ms before aborting
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp (hunk 3 above)
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp]:
```suggestion
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp [hunk 5]
```diff
@@ -283,9 +283,9 @@ void FiniteDifferenceSolver::MacroscopicEvolveM_2nd (
...
-        int M_max_iter = 100;
+        int M_max_iter = macroscopic_properties->getmag_max_iter();
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp:286]:
```suggestion
        // maximum number of iterations allowed
        int M_max_iter = macroscopic_properties->getmag_max_iter();
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp [hunk 6]
```diff
@@ -283,9 +283,9 @@ void FiniteDifferenceSolver::MacroscopicEvolveM_2nd (
...
         int M_iter = 0;
-        amrex::Real M_tol = 0.0001;
+        amrex::Real M_tol = macroscopic_properties->getmag_tol();
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp:289]:
```suggestion
        // relative tolerance stopping criteria for 2nd-order iterative algorithm
        amrex::Real M_tol = macroscopic_properties->getmag_tol();
```

**Review by ajnonaka — APPROVED** [review]

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp [hunk 7]
```diff
@@ -73,6 +73,16 @@ MacroscopicProperties::ReadParameters ()
...
+
+    m_mag_normalized_error = 0.1;
+    pp.get("mag_normalized_error",m_mag_normalized_error);
+
+    m_mag_max_iter = 100;
+    pp.get("mag_max_iter",m_mag_max_iter);
+
+    m_mag_tol = 0.0001;
+    pp.get("mag_tol",m_mag_tol);
+
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp:85]: change these pp.get to pp.query
get is required - if the argument is not found a run-time error is generated
query is optional.

**Review by RevathiJambunathan — APPROVED** [review]
Thanks for the PR, @ruohai0925 and the review @ajnonaka 
Looks good to me!

**Pull request merged**